File: src/application/excel/MonsterData.java
package application.excel;

import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

import org.apache.poi.xssf.usermodel.XSSFCellStyle;
import org.apache.poi.xssf.usermodel.XSSFSheet;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;

import application.controllers.MonsterEditor;
import application.models.AbstractItem;
import application.models.Monster;
import application.models.lists.AbstractItems;
import application.models.lists.DropTables;
import application.models.lists.Lists;
import application.models.lists.Monsters;
import application.models.lists.Spells;

public class MonsterData
{
    public static void makeChangeList()
    {
	try (XSSFWorkbook workbook = new XSSFWorkbook(); OutputStream fileOut = Excel.createOutputStream("Monster.xlsx"))
	{
	    XSSFSheet sheet = workbook.createSheet("Monster Data");
	    XSSFCellStyle defaultStyle = Excel.setDefaultStyle(workbook);
	    XSSFCellStyle boldStyle = Excel.setBoldStyle(workbook);

	    List<RowData> rowData = new ArrayList<RowData>();

	    Monsters monsters = new Monsters(Lists.getMonsters(), false);

	    if (MonsterEditor.isChronologicalOrderSelected())
	    {
		monsters.chronologicalIndexSort();
	    }

	    rowData.add(new RowData(boldStyle, new String[]
	    {
		    "Stats", "", "Resists", "", "Spells", "", "Drops"
	    }));
	    rowData.add(new RowData(boldStyle, new String[]
	    {
		    ""
	    }));

	    Spells spells = new Spells(Lists.getSpells(), true);
	    DropTables dropTables = new DropTables(Lists.getDropTables());
	    AbstractItems<AbstractItem> abstractItems = new AbstractItems<>(Lists.getAbstractItems(), true);

	    for (int i = 0; i < monsters.getModels().size(); i++)
	    {
		Monster m = monsters.getModels().get(i);
		Monster Dm = monsters.getDModel(m);

		List<String> spellNames = spells.getNames(m.getSpells());
		List<String> DspellNames = spells.getDNames(Dm.getSpells());

		List<String> spellChances = m.getSpellChanceStrings();
		List<String> DspellChances = Dm.getSpellChanceStrings();

		List<String> itemDropNames = abstractItems.getNames(dropTables.getModels().get(m.getItemDropSet()));
		List<String> DitemDropNames = abstractItems.getDNames(dropTables.getDModels().get(Dm.getItemDropSet()));

		int index = 0;
		rowData.add(new RowData(defaultStyle, new String[]
		{
			m.getName()
		}));

		rowData.add(new RowData(defaultStyle, new String[]
		{
			"HP", Excel.CompareValues(Dm.getHp(), m.getHp()), "Boss Bit", Excel.CompareValues(Dm.getRunFlag(), m.getRunFlag()),
			Excel.CompareValues(DspellNames.get(index), spellNames.get(index)), Excel.CompareValues(DspellChances.get(index), spellChances.get(index)),
			Excel.CompareValues(Excel.Pget(DitemDropNames, index), Excel.Pget(itemDropNames, index++))
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			"MP", Excel.CompareValues(Dm.getMp(), m.getMp()), "Lightning", Excel.CompareValues(Dm.getLaserRes(), m.getLaserRes()),
			Excel.CompareValues(DspellNames.get(index), spellNames.get(index)), Excel.CompareValues(DspellChances.get(index), spellChances.get(index)),
			Excel.CompareValues(Excel.Pget(DitemDropNames, index), Excel.Pget(itemDropNames, index++))
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Power", Excel.CompareValues(Dm.getPower(), m.getPower()), "Unknown", Excel.CompareValues(Dm.getUnknownRes1(), m.getUnknownRes1()),
			Excel.CompareValues(DspellNames.get(index), spellNames.get(index)), Excel.CompareValues(DspellChances.get(index), spellChances.get(index)),
			Excel.CompareValues(Excel.Pget(DitemDropNames, index), Excel.Pget(itemDropNames, index++))
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Guard", Excel.CompareValues(Dm.getGuard(), m.getGuard()), "Unknown2", Excel.CompareValues(Dm.getUnknownRes2(), m.getUnknownRes2()),
			Excel.CompareValues(DspellNames.get(index), spellNames.get(index)), Excel.CompareValues(DspellChances.get(index), spellChances.get(index)),
			Excel.CompareValues(Excel.Pget(DitemDropNames, index), Excel.Pget(itemDropNames, index++))
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Magic", Excel.CompareValues(Dm.getMagic(), m.getMagic()), "Fire", Excel.CompareValues(Dm.getFireRes(), m.getFireRes()),
			Excel.CompareValues(DspellNames.get(index), spellNames.get(index)), Excel.CompareValues(DspellChances.get(index), spellChances.get(index)),
			Excel.CompareValues(Excel.Pget(DitemDropNames, index), Excel.Pget(itemDropNames, index++))
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Speed", Excel.CompareValues(Dm.getSpeed(), m.getSpeed()), "Ice", Excel.CompareValues(Dm.getIceRes(), m.getIceRes()),
			Excel.CompareValues(DspellNames.get(index), spellNames.get(index)), Excel.CompareValues(DspellChances.get(index), spellChances.get(index)),
			Excel.CompareValues(Excel.Pget(DitemDropNames, index), Excel.Pget(itemDropNames, index++))
		}));

		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Gold", Excel.CompareValues(Dm.getGold(), m.getGold()), "Vacuum", Excel.CompareValues(Dm.getVacuumRes(), m.getVacuumRes()),
			Excel.CompareValues(DspellNames.get(index), spellNames.get(index)), Excel.CompareValues(DspellChances.get(index), spellChances.get(index)),
			Excel.CompareValues(Excel.Pget(DitemDropNames, index), Excel.Pget(itemDropNames, index++))
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			"Exp", Excel.CompareValues(Dm.getExperience(), m.getExperience()), "Debuff", Excel.CompareValues(Dm.getDebuffRes(), m.getDebuffRes()),
			Excel.CompareValues(DspellNames.get(index), spellNames.get(index)), Excel.CompareValues(DspellChances.get(index), spellChances.get(index)),
			Excel.CompareValues(Excel.Pget(DitemDropNames, index), Excel.Pget(itemDropNames, index++))
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			""
		}));
		rowData.add(new RowData(defaultStyle, new String[]
		{
			""
		}));

	    }
	    Excel.writeDataToSheet(sheet, rowData);
	    workbook.write(fileOut);
	}

	catch (IOException e)
	{
	    e.printStackTrace();
	}
    }
}

File: src/application/excel/MonsterData.kt
package application.excel

import application.ROM
import application.controllers.MonsterEditor
import application.models.lists.*

class MonsterData(fileName: String) : Excel(fileName)
{
	override fun makeChangeList()
	{
		val sheet = workbook.createSheet("Monster Data")
		val sheetData = SheetData()
		//val monsters = Monsters(Lists.monsters, false)
		var monsters = Monsters(Lists.monsters, ROM.settings.showEmptyValues, ROM.settings.showMonsterDuplicates)

		if (MonsterEditor.isChronologicalOrderSelected) monsters.chronologicalIndexSort()

		sheetData.addRow(RowData(boldStyle, mutableListOf("Stats", "", "Resists", "", "Spells", "", "Drops")))
		sheetData.addRow(RowData(boldStyle, mutableListOf("")))

		val spells = Spells(Lists.spells, true)
		val dropTables = DropTables(Lists.dropTables)
		val abstractItems = AbstractItems(Lists.abstractItems, true)

		for (i in 0..monsters.models.size - 1)
		{
			val m = monsters.models[i]
			val Dm = monsters.getDModel(m)

			val spellNames = spells.getNames(m.spells)
			val DspellNames = spells.getDNames(Dm.spells)

			val spellChances = m.spellChanceStrings
			val DspellChances = Dm.spellChanceStrings

			val itemDropNames = abstractItems.getNames(dropTables.models[m.itemDropSet])
			val DitemDropNames = abstractItems.getDNames(dropTables.dModels[Dm.itemDropSet])

			var index = 0
			sheetData.addRow(RowData(boldStyle, mutableListOf(m.name)))

			sheetData.addRow(RowData(defaultStyle, mutableListOf("HP", CompareValues(Dm.hp, m.hp), "Boss Bit", CompareValues(Integer.toHexString(Dm.runFlag), Integer.toHexString(m.runFlag)), CompareValues(DspellNames[index], spellNames[index]), CompareValues(DspellChances[index], spellChances[index]), CompareValues(Pget(DitemDropNames, index), Pget(itemDropNames, index++)))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("MP", CompareValues(Dm.mp, m.mp), "Lightning", CompareValues(Dm.laserRes, m.laserRes), CompareValues(DspellNames[index], spellNames[index]), CompareValues(DspellChances[index], spellChances[index]), CompareValues(Pget(DitemDropNames, index), Pget(itemDropNames, index++)))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("Power", CompareValues(Dm.power, m.power), "Unknown", CompareValues(Dm.unknownRes1, m.unknownRes1), CompareValues(DspellNames[index], spellNames[index]), CompareValues(DspellChances[index], spellChances[index]), CompareValues(Pget(DitemDropNames, index), Pget(itemDropNames, index++)))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("Guard", CompareValues(Dm.guard, m.guard), "Unknown2", CompareValues(Dm.unknownRes2, m.unknownRes2), CompareValues(DspellNames[index], spellNames[index]), CompareValues(DspellChances[index], spellChances[index]), CompareValues(Pget(DitemDropNames, index), Pget(itemDropNames, index++)))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("Magic", CompareValues(Dm.magic, m.magic), "Fire", CompareValues(Dm.fireRes, m.fireRes), CompareValues(DspellNames[index], spellNames[index]), CompareValues(DspellChances[index], spellChances[index]), CompareValues(Pget(DitemDropNames, index), Pget(itemDropNames, index++)))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("Speed", CompareValues(Dm.speed, m.speed), "Ice", CompareValues(Dm.iceRes, m.iceRes), CompareValues(DspellNames[index], spellNames[index]), CompareValues(DspellChances[index], spellChances[index]), CompareValues(Pget(DitemDropNames, index), Pget(itemDropNames, index++)))))

			sheetData.addRow(RowData(defaultStyle, mutableListOf("Gold", CompareValues(Dm.gold, m.gold), "Vacuum", CompareValues(Dm.vacuumRes, m.vacuumRes), CompareValues(DspellNames[index], spellNames[index]), CompareValues(DspellChances[index], spellChances[index]), CompareValues(Pget(DitemDropNames, index), Pget(itemDropNames, index++)))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("Exp", CompareValues(Dm.experience, m.experience), "Debuff", CompareValues(Dm.debuffRes, m.debuffRes), CompareValues(DspellNames[index], spellNames[index]), CompareValues(DspellChances[index], spellChances[index]), CompareValues(Pget(DitemDropNames, index), Pget(itemDropNames, index++)))))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("")))
			sheetData.addRow(RowData(defaultStyle, mutableListOf("")))
		}
		sheet.createFreezePane(0,2,0,2)
		writeDataToSheet(sheet, sheetData)
	}
}


